packages/api/src/Base.ts

```diff
@@ -0,0 +1,216 @@
+   * ```javascript
+   * import Api from '@polkadot/api/rx';
+   *
+   * new Api().isReady.subscribe((api) => {
+   *   api.rpc.newHead().subscribe((header) => {
+   *     console.log(`new block #${header.blockNumber.toNumber()}`);
+   *   });
+   * });
+   * ```
+   */
+  constructor (wsProvider?: WsProvider) {
+    super();
+
+    this._rpcBase = new RpcRx(wsProvider);
+    this._rpc = this.decorateRpc(this._rpcBase);
+
+    this.init();
+  }
+
+  /**
+   * @description Contains the genesis Hash of the attached chain. Apart from being useful to determine the actual chain, it can also be used to sign immortal transactions.
+   */
+  get genesisHash (): Hash {
+    assert(!isUndefined(this._genesisHash), INIT_ERROR);
+
+    return this._genesisHash as Hash;
+  }
+
+  /**
+   * @description Yields the current attached runtime metadata. Generally this is only used to construct extrinsics & storage, but is useful for current runtime inspection.
+   */
+  get runtimeMetadata (): RuntimeMetadata {
+    assert(!isUndefined(this._runtimeMetadata), INIT_ERROR);
+
+    return this._runtimeMetadata as RuntimeMetadata;
+  }
+
+  /**
+   * @description Contains the version information for the current runtime.
+   */
+  get runtimeVersion (): RuntimeVersion {
+    assert(!isUndefined(this._runtimeVersion), INIT_ERROR);
+
+    return this._runtimeVersion as RuntimeVersion;
+  }
+
+  /**
+   * @description Contains all the raw rpc sections and their subsequent methods in the API as defined by the jsonrpc interface definitions.
+   * @example
+   * <BR>
+   *
+   * ```javascript
+   * api.rpc.chain
+   *   .newHead()
+   *   .subscribe((header) => {
+   *     console.log('new header', header);
+   *   });
+   * ```
+   */
+  get rpc (): R {
+    return this._rpc;
+  }
+
+  /**
+   * @description Contains all the chain state modules and their subsequent methods in the API. These are attached dynamically from the runtime metadata.
+   * @example
+   * <BR>
+   *
+   * ```javascript
+   * api.st.balances
+   *   .freeBalance(<accountId>)
+   *   .subscribe((balance) => {
+   *     console.log('new balance', balance);
+   *   });
+   * ```
+   */
+  get st (): S {
+    assert(!isUndefined(this._storage), INIT_ERROR);
+
+    return this._storage as S;
+  }
+
+  /**
+   * @description Contains all the extrinsic modules and their subsequent methods in the API. It allows for the construction of transactions and the submission thereof. These are attached dynamically from the runtime metadata.
+   * @example
+   * <BR>
+   *
+   * ```javascript
+   * api.tx.balances
+   *   .transfer(<recipientId>, <balance>)
+   *   .sign(<keyPair>, <accountNonce>, <blockHash (optional)>)
+   *   .send()
+   *   .subscribe((status) => {
+   *     console.log('tx status', status);
+   *   });
+   * ```
+   */
+  get tx (): T {
+    assert(!isUndefined(this._extrinsics), INIT_ERROR);
+
+    return this._extrinsics as T;
+  }
+
+  private init (): void {
+    let isReady: boolean = false;
```

Comment:
Is this redundant to the isReady property? One idea is to have isReady as an Observable on this abstract class, and in the Promise implementation, add `get isReady() { return this._isReady.toPromise() }`

Reply:
I went down that route, I think it is easier to just emit the 'ready' and the extention can do what it wants. Will have to see with promise, maybe you are right here, I'm not 100% either way.